src/openrct2/core/String.cpp

```diff
@@ -517,4 +518,97 @@ namespace String
         size_t stringLength = endSubstr - startSubstr + 1;
         return std::string(startSubstr, stringLength);
     }
+
+
+    size_t ToUpper(utf8 * buffer, size_t bufferSize, const utf8 * src)
+    {
+        size_t remainingSize = bufferSize;
+
+        while (*src != '\0' && remainingSize)
+        {
+            size_t numBytes;
+            if ((*src & 0x80) == 0)
+            {
+                // 1-byte characters can just be treated as ASCII
+                *buffer = toupper(*src);
+                buffer++;
+                remainingSize--;
+                numBytes = 1;
+            }
+            else
+            {
+                uint32 word;
+                uint32 upperWord;
+                if ((*src & 0xF0) == 0xF0)
+                {
+                    word = ((uint8)src[0] << 24) + ((uint8)src[1] << 16) + ((uint8)src[2] << 8) + (uint8)src[3];
```

Comment:
An invalid UTF-8 string could cause an out-of-bounds access here, is this a possible input?  I assumed not, but if it's a possible situation, I should definitely add a null check for each index.